test/router/V4Router.t.sol

```diff
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {IV4Router} from "../../src/interfaces/IV4Router.sol";
+import {RoutingTestHelpers} from "../shared/RoutingTestHelpers.sol";
+import {Plan, ActionsRouterPlanner} from "../shared/ActionsRouterPlanner.sol";
+import {Actions} from "../../src/libraries/Actions.sol";
+
+contract V4RouterTest is RoutingTestHelpers {
+    using CurrencyLibrary for Currency;
+    using ActionsRouterPlanner for Plan;
+
+    function setUp() public {
+        setupRouterCurrenciesAndPoolsWithLiquidity();
+        plan = ActionsRouterPlanner.init();
+    }
+
+    function test_swapExactInputSingle_zeroForOne() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
+
+        IV4Router.ExactInputSingleParams memory params =
+            IV4Router.ExactInputSingleParams(key0, true, uint128(amountIn), 0, 0, bytes(""));
+
+        uint256 prevBalance0 = key0.currency0.balanceOf(address(this));
+        uint256 prevBalance1 = key0.currency1.balanceOf(address(this));
+
+        plan = plan.add(Actions.SWAP_EXACT_IN_SINGLE, abi.encode(params));
+        _finalizePlan(key0.currency0, key0.currency1, address(this));
+        bytes memory data = plan.encode();
+
+        router.executeActions(data);
+
+        uint256 newBalance0 = key0.currency0.balanceOf(address(this));
+        uint256 newBalance1 = key0.currency1.balanceOf(address(this));
+
+        assertEq(prevBalance0 - newBalance0, amountIn);
+        assertEq(newBalance1 - prevBalance1, expectedAmountOut);
+    }
+
+    function test_swapExactInputSingle_oneForZero() public {
+        uint256 amountIn = 1 ether;
+        uint256 expectedAmountOut = 992054607780215625;
```

Comment:
where is this number from? I think you could use deltas instead of hardcoding?

Reply:
gonna do this in the next PR - #221